docs/advanced/misc.rst

```diff
@@ -304,3 +304,33 @@ the default settings are restored to prevent unwanted side effects.
 
 .. [#f4] http://www.sphinx-doc.org
 .. [#f5] http://github.com/pybind/python_example
+
+.. _avoiding-cpp-types-in-docstrings:
+
+Avoiding C++ types in docstrings
+================================
+
+Docstrings are generated at the time of the declaration, e.g. when ``.def(...)`` is called.
+At this point argument and return types should be known to pybind11.
+If a custom type is not exposed yet through a ``py::class_`` constructor or a custom type caster,
+its C++ type name will be used instead to generate the signature in the docstring:
+
+.. code-block:: text
+
+     |  __init__(...)
+     |      __init__(self: example.Foo, arg0: ns::Bar) -> None
+                                              ^^^^^^^
+
+
+This limitation can be circumvented by "forward declaring" bound classes:
```

Comment:
I don't like the use of "forward declaration" here. How about

> ...by separating the bindings of the class itself from the bindings of the respective class' contents:

It's just that "forward declaration" has a very specific meaning in C++ and this isn't talking about that.

Reply:
That's why it's in quotes. Quotes makes reader to wonder what that means and turn to example to understand. I don't think analogy to _forward declaration in C++_ creates a confusion to reader. 

I had to re-read the suggested replacement to understand what it means (because "bindings of the class" at first read includes "class content bindings" as well). It takes me 5 more seconds to think about before I look at example and finally understand.

While you are formally correct, I think it's better to keep it close to current form.

Reply:
How about instead

> ... by ensuring that C++ classes are registered with pybind11 before they are used as a parameter to a function

Reply:
I don't think quotes is enough of a hint for "nevermind terminology, look at the example". I know my suggestion isn't perfect either.

@virtuald's suggestion reads better in my opinion, but I would be happy to hear other opinions.

Reply:
I like @virtuald's suggestion.

Reply:
I had noticed this, but couldn't immediately think of a better way of phrasing it either. So I let this go and thought the analogy does kind of make sense.

But @bstaletic's or @virtuald's also both sound good to me. The main problem is we don't seem to have a fixed terminology.

To match the docs [here](https://pybind11.readthedocs.io/en/stable/classes.html), and combine all suggestions, what about:

> ... by creating the class bindings before creating bindings for functions or methods with a signature that contains the corresponding C++ class:

or (but this is probably too long)

> ... by creating the class bindings (and thus registering them with pybind11) before creating bindings for functions or methods with a signature that contains the corresponding C++ class:

I guess the "forward declaration" becomes clear from the example, then?
Potentially, we could add a method binding that isn't just the constructor? And/or a binding for a free-standing function?

Reply:
(Kind of [relevant xkcd](https://xkcd.com/927/)? "Here's another suggestion.")

Reply:
I think the goal is to make it easy to read. The phrase "class bindings" usually includes binding of class members/functions in pybind11 docs. Therefore I find wordinig that use it more confusing than those that doesn't. 

I've update the PR with initial @virtuald's wording which seems to fine to everybody.

Reply:
I was slightly afraid that "registering with pybind11" would not be clear to users, but if it's fine with @bstaletic, fine with me as well.

Also, to be really pedantic (we're here anyway :smile:): "as a parameter or return type of a function" ? I promise that's my last remark/request, and that I'll press "merge" after @bstaletic then approves ;-)

Reply:
Ok, I've updated the PR. 
(Please squash all the commits on merge to avoid history pollution)

Reply:
Thanks!
And jep, that's pybind11's default, these days :-)

Reply:
@bstaletic: last call. We're all good now?